core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/UpgradeService.java

```diff
@@ -0,0 +1,129 @@
+import com.sequenceiq.cloudbreak.service.image.ImageCatalogService;
+import com.sequenceiq.cloudbreak.service.image.ImageService;
+import com.sequenceiq.cloudbreak.service.image.StatedImage;
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+import com.sequenceiq.cloudbreak.workspace.model.User;
+import com.sequenceiq.distrox.api.v1.distrox.endpoint.DistroXV1Endpoint;
+
+@Component
+public class UpgradeService {
+
+    private static final Set<Status> UPGRADEABLE_ATTACHED_DISTRO_X_STATES = Sets.immutableEnumSet(Status.STOPPED, Status.DELETE_COMPLETED);
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private ImageService imageService;
+
+    @Inject
+    private ImageCatalogService imageCatalogService;
+
+    @Inject
+    private ComponentConfigProviderService componentConfigProviderService;
+
+    @Inject
+    private ClusterService clusterService;
+
+    @Inject
+    private DistroXV1Endpoint distroXV1Endpoint;
+
+    @Inject
+    private TransactionService transactionService;
+
+    public UpgradeOption getUpgradeOptionByName(Long workspaceId, String name, User user) {
+        try {
+            return transactionService.required(() -> {
+                Optional<Stack> stack = stackService.findStackByNameAndWorkspaceId(name, workspaceId);
+                if (stack.isPresent()) {
+                    return getUpgradeOption(stack.get(), workspaceId, user);
+                } else {
+                    throw notFoundException("Stack", name);
+                }
+            });
+        } catch (TransactionExecutionException e) {
+            throw new TransactionRuntimeExecutionException(e);
+        }
+    }
+
+    private UpgradeOption getUpgradeOption(Stack stack, Long workspaceId, User user) {
+        if (clusterService.repairSupported(stack) && attachedClustersStoppedOrDeleted(stack)) {
+            try {
+                Blueprint blueprint = stack.getCluster().getBlueprint();
+                Image image = componentConfigProviderService.getImage(stack.getId());
+                ImageSettingsV4Request imageSettingsV4Request = toImageSettingsRequest(image);
+                boolean baseImage = useBaseImage(image);
+                StatedImage latestImage = imageService
+                        .determineImageFromCatalog(workspaceId, imageSettingsV4Request, stack.getCloudPlatform().toLowerCase(), blueprint, baseImage, user);
+                if (!Objects.equals(image.getImageId(), latestImage.getImage().getUuid())) {
+                    return new UpgradeOption(latestImage.getImage().getUuid(), latestImage.getImageCatalogName());
+                } else {
+                    return new UpgradeOption();
+                }
+            } catch (CloudbreakImageNotFoundException | CloudbreakImageCatalogException e) {
+                throw new BadRequestException(e.getMessage());
+            }
+        } else {
+            return new UpgradeOption();
```

Comment:
Maybe we should give a different HTTP code if there is no new image and we could use Optional here instead of an empty object